lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetMetadataUtils.java

```diff
@@ -0,0 +1,212 @@
+ */
+package io.trino.parquet.writer;
+
+import org.apache.parquet.column.statistics.BinaryStatistics;
+import org.apache.parquet.format.Statistics;
+import org.apache.parquet.internal.column.columnindex.BinaryTruncator;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.schema.ColumnOrder;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+// Based on org.apache.parquet.format.converter.ParquetMetadataConverter
+public class ParquetMetadataUtils
+{
+    private ParquetMetadataUtils() {}
+
+    public static <T extends Comparable<T>> Statistics toParquetStatistics(org.apache.parquet.column.statistics.Statistics<T> stats, int truncateLength)
+    {
+        Statistics formatStats = new Statistics();
+        // We skip the `withinLimit(stats, truncateLength)` check from equivalent parquet-mr code
+        // which avoids writing stats altogether if they're above 4K in size.
+        // This is to allow readers to take advantage of truncated stats for filtering rather than having no stats.
+        if (!stats.isEmpty()) {
+            formatStats.setNull_count(stats.getNumNulls());
+            if (stats.hasNonNullValue()) {
+                byte[] min;
+                byte[] max;
+
+                if (stats instanceof BinaryStatistics && truncateLength != Integer.MAX_VALUE) {
+                    BinaryTruncator truncator = BinaryTruncator.getTruncator(stats.type());
+                    min = truncateMin(truncator, truncateLength, stats.getMinBytes());
+                    max = truncateMax(truncator, truncateLength, stats.getMaxBytes());
+                }
+                else {
+                    min = stats.getMinBytes();
+                    max = stats.getMaxBytes();
+                }
+                // Fill the former min-max statistics only if the comparison logic is
+                // signed so the logic of V1 and V2 stats are the same (which is
+                // trivially true for equal min-max values)
+                if (sortOrder(stats.type()) == SortOrder.SIGNED || Arrays.equals(min, max)) {
+                    formatStats.setMin(min);
+                    formatStats.setMax(max);
+                }
+
+                if (isMinMaxStatsSupported(stats.type()) || Arrays.equals(min, max)) {
+                    formatStats.setMin_value(min);
+                    formatStats.setMax_value(max);
+                }
+            }
+        }
+        return formatStats;
+    }
+
+    private static byte[] truncateMin(BinaryTruncator truncator, int truncateLength, byte[] input)
+    {
+        return truncator.truncateMin(Binary.fromConstantByteArray(input), truncateLength).getBytes();
+    }
+
+    private static byte[] truncateMax(BinaryTruncator truncator, int truncateLength, byte[] input)
+    {
+        return truncator.truncateMax(Binary.fromConstantByteArray(input), truncateLength).getBytes();
```

Comment:
We need `NO_OP_TRUNCATOR` for `INT96`

Reply:
> the call looks thread-unsafe because of shared buffer use in validator.checkValidity(buffer)
Good point, the usage of `CharsetValidator#dummyBuffer` does look thread unsafe
I think we need to copy paste BinaryTruncator too to resolve it :(

Reply:
or synchronize the call on our side (on a class instance)

how much contention we can expect here vs how much it will cost us to  copy & maintain even more code?

Reply:
I've updated this to avoid copying code now.
I'm taking a lock only when truncation is possible, which should not be a very common scenario.
Typically creating metadata for the footer is a fast operation, majority of time in parquet writer is spent in other places (compression, encoding, write to storage)